### DISCUSSION TITLE
Semantic Kernel plugin templates from database

### DISCUSSION THREAD

**Discussion by jamesmkfoo23** [microsoft/semantic-kernel#7106] [Q&A] [answered]
Is it possible to store plugin templates in database and create plugin from retrieved .json settings and .txt with input parameters denoted by $ and {{ function call }}?

**Comment by dmytrostruk** [comment, marked as answer]
@jamesmkfoo23 In theory it's possible, as well as storing plugin templates in files. But you will have to create instances of these functions on your side, in similar way like we do this with files:
https://github.com/microsoft/semantic-kernel/blob/a0a753abe9dae502866d7c5e2fb263f2881ce4cc/dotnet/src/SemanticKernel.Core/KernelExtensions.cs#L517

It would be interesting to learn about your use-case to recommend the best approach how to work with plugin templates, so if you could share any details, that would be great! Thank you!

**Comment by jamesmkfoo23** [comment]
Hi @dmytrostruk , I am using Python kernel.add_plugin at the moment. 

  ```
kernel.add_plugin(
            parent_directory=plugins_directory,
            plugin_name="PluginName",
        )
```

What I am thinking of is to define the plugins which consist of prompt template & config json, in the DB. 

From Front End application to get the plugin information and parameterize the inputs for submission, the prompt preview is displayed for the user which explains how the inputs are used in the prompt. Think of a use case which is more toward a specific utility or task that could use LLM/tools.

Being able to store the plugins in DB can allow for better versioning and test result tracking, but I would like to avoid defining it twice:
i) in the DB 
ii) in the SK prompt file.

**Reply by dmytrostruk** [reply]
@jamesmkfoo23 Thanks for provided information! Based on your case, I think it's achievable. You can store your prompt template and config.json in database as strings, then you initialize your function like this:

Create `PromptTemplateConfig` object from config.json string:
```python
prompt_template_config = PromptTemplateConfig.from_json("config.json from DB")
```

Set template in `PromptTemplateConfig` object:
```python
prompt_template_config.template = "prompt template from DB"
```

Then you can initialize your function and pass prompt template config to it, like in this example:
https://github.com/microsoft/semantic-kernel/blob/79207ffbbc6655391dd072b0afa994de407b6be0/python/samples/concepts/prompt_templates/configuring_prompts.py#L42-L46

Please let us know if this helps to achieve your scenario. Thanks!
